tools/gsnapshot/Operations.cs

```diff
@@ -0,0 +1,204 @@
+/*
+   Copyright 2022 Google LLC
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+using Microsoft.Extensions.Logging;
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Compute.v1;
+using Google.Apis.CloudResourceManager.v1;
+using Google.Apis.Services;
+using Newtonsoft.Json;
+using ByteSizeLib;
+
+using ComputeData = Google.Apis.Compute.v1.Data;
+using CrmData = Google.Apis.CloudResourceManager.v1.Data;
+
+#nullable enable
+#pragma warning disable CS1998
+namespace GSnapshot
+{
+    public class Operations
+    {
+        private readonly ILogger<Runner> _logger;
+
+        private ComputeService computeService = new ComputeService(new BaseClientService.Initializer
+        {
+            HttpClientInitializer = Utils.GetCredential(),
+            ApplicationName = Utils.ApplicationName,
+        });
+
+        public Dictionary<string, string> AllRegions = new Dictionary<string, string>();
+        public Dictionary<string, List<string>> AllZones = new Dictionary<string, List<string>>();
+        public Dictionary<string, ComputeData.Instance> AllInstances = new Dictionary<string, ComputeData.Instance>();
+
+
+        public Operations(ILogger<Runner> logger)
+        {
+            _logger = logger;
+        }
+
+        // Wait for long running operation to complete.
+        public async Task<bool> WaitForOperation(string projectId, string type, string location, string id)
```

Comment:
This method and the next one have a lot of duplicated code to separate between global, regional and zonal. Maybe you could consider separating the case specific code and the "request done" checking. And use Func delegates to reduce the amount of duplication/boilerplate.

Reply:
Separated response processing, since that's common.